### PR TITLE
GPULightmapper exclude back-face triangles while calculating bounces

### LINKED ISSUE #51638: GPULightmapper leaks light in situations where the CPU lightmapper wouldn't
### Godot version

v4.0.dev.calinou [6f043f7a1]

### System information

Mx Linux (Debian Based)

### Issue description

The lightmapper leaks even with thick walls.

image

### Steps to reproduce

Use the minimal reproduction project and bake the lightmap

### Minimal reproduction project

[GPULightmapLeak.zip](https://github.com/godotengine/godot/files/6984512/GPULightmapLeak.zip)

### PR DESCRIPTION
**Pull request by williamd67**
Edges that are at the edge of a plane, may get behind the scene and will hit
back-face triangles which where included in the lighting calculations. This
caused leaking of light at the edge of planes.

In case a ray hits back-face triangle, it is skipped in the bounce calculations.

Now it should fix #51638.

### CODE REVIEW

**Comment by Calinou** [comment]
Does this close https://github.com/godotengine/godot/issues/51638?

**Comment by williamd67** [comment]
> Does this close  ##51638?

It did not remove it completely although together with increasing the setting for `Lightmap Scale` of the `Global Illumination` from `1x` to `2x`. It did solve it. I plan to investigate it a little further.

**Comment by williamd67** [comment]
I found an issue in the implementation of the exclusion. I will recommit an improved version of this fix.

**Comment by jcostello** [comment]
I imagine that this PR fix this, right @williamd67? The house is empty inside with the boards with only one side

image

File: modules/lightmapper_rd/lm_compute.glsl [hunk 1]
```diff
@@ -118,13 +118,17 @@ bool ray_hits_triangle(vec3 from, vec3 dir, float max_dist, vec3 p0, vec3 p1, ve
 bool trace_ray(vec3 p_from, vec3 p_to
 #if defined(MODE_BOUNCE_LIGHT) || defined(MODE_LIGHT_PROBES)
 		,
-		out uint r_triangle, out vec3 r_barycentric
+		out uint r_triangle, out vec3 r_barycentric, out bool backface
```
Comment (akien-mga) [inline modules/lightmapper_rd/lm_compute.glsl:121]: By convention, parameters which are modified by the function to return values use the `r_` prefix (like the two other `out` parameters here).

Comment (williamd67) [reply modules/lightmapper_rd/lm_compute.glsl]: Thanks for the feedback, I updated the PR and removed the extra parameter.

**Comment by williamd67** [comment]
I fixed the issue 'distance checking for back-faces was skipped' I mentioned before.

Now it should fix #51638.

**Review by JFonS — APPROVED** [review]
Changes look great, awesome work :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**